azurelinuxagent/common/event.py

```diff
@@ -342,12 +335,36 @@ def add_metric(self, category, counter, instance, value, log_event=False):
         event.parameters.append(TelemetryEventParam('Instance', instance))
         event.parameters.append(TelemetryEventParam('Value', value))
 
+        self._add_default_parameters_to_event(event)
         data = get_properties(event)
         try:
             self.save_event(json.dumps(data))
         except EventError as e:
             logger.error("{0}", e)
 
+    @staticmethod
+    def _add_default_parameters_to_event(event):
+        # We write the GAVersion here rather than add it in azurelinuxagent.ga.monitor.MonitorHandler.add_sysinfo
+        # as there could be a possibility of events being sent with newer version of the agent, rather than the agent
+        # version generating the event.
+        # Old behavior example: V1 writes the event on the disk and finds an update immediately, and updates. Now the
+        # new monitor thread would pick up the events from the disk and send it with the CURRENT_AGENT, which would have
+        # newer version of the agent. This causes confusion.
+        event.parameters.append(TelemetryEventParam("GAVersion", CURRENT_AGENT))
+
+        # ContainerId can change due to live migration and we want to preserve the container Id of the container writing
+        # the event, rather than sending the event.
+        event.parameters.append(TelemetryEventParam('ContainerId', get_container_id_from_env()))
+
+        # This is used as the actual time of event generation.
+        event.parameters.append(TelemetryEventParam('OpcodeName', datetime.utcnow().__str__()))
```

Comment:
I wanted to do it, but it is a lot of change across 

In event.py, we have `from datetime import datetime`, where as in monitor.py, we only do `import datetime`. I was thinking of changing it but it would be a lot of change for this PR, which is why I kept it as it is. Will do that subsequently.